src/debug.rs

```diff
@@ -52,6 +52,65 @@ pub fn dump_memory(start_address: *const usize, count: isize) {
     }
 }
 
+/// Use the LowLevelDebug capsule (if present) to print a single number. If the
+/// capsule is not present, this is a no-op.
+#[cfg(target_arch = "arm")]
+#[inline(always)] // Improve reliability for relocation issues
+pub fn low_level_print1(value: usize) {
+    unsafe {
+        asm!("svc 2"
```

Comment:
Is there a reason to use raw assembly instead of relying on `syscall::command`, which already supports all architectures, and has [mock implementations](https://github.com/tock/libtock-rs/blob/master/src/syscalls_mock.rs)? Duplicating the amount of unsafe assembly doesn't seem reasonable to me from a safety perspective. It's also extra lines of code.

Also, why are the output operands and clobbers different than in `syscall.rs`?

Reply:
The clobbers are different because `syscalls::command` must work with arbitrary drivers, whereas these methods only need to work with LowLevelDebug. Unlike many drivers, LowLevelDebug never mutates application memory, so we do not need the memory clobber here. The decision to mark `r0` as a clobber rather than a return value was fairly arbitrary; the return value from LowLevelDebug should be ignored either way.

There are a few reasons I used raw assembly:

1. I want the call inlined directly in the calling code, in case function calls are broken.
2. It will produce simpler assembly due to the reduced clobbers. This will help debugging, as you'd presumably be reading the assembly directly while using this functionality.

As an alternative, I could put the `#[inline(always)]` annotation on `syscalls::command` and call that. That addresses issue 1 but not issue 2.

What do you think?

Reply:
I don't have an opinion on how libtock-rs moves forward with this architecturally, but just wanted to chime in quickly and say that skimming this caused me to think for a few minutes about the limited clobber list for this case, and that I'll confirm this will work correctly.

Reply:
1. Regarding 1, it would be quite unlikely that the compiler doesn't inline such a trivial function when optimizations are enabled. But regular syscalls should also be inlined, so adding `#[inline(always)]` on all syscalls makes sense.

2. Regarding 2, here is what godbolt.org has to say (compiled in optimized mode): https://godbolt.org/z/JCWjzs
At least in simple cases, removing the memory clobber doesn't change the assembly. Also note that the syscall function was inlined.

    More generally, shouldn't the full memory clobber act as a barrier preventing the compiler from re-ordering other operations before/after the debug syscall?

    In that sense, it seems that the memory barrier is actually better: otherwise, an operation appearing after the debug syscall in the code but reordered before it by the compiler could mess up the manual debugging.

    Is such reordering what one would expect when doing such "manual printf" debugging? And if after reordering operations the compiler also apply some optimizations, wouldn't the produced assembly be more complex, due to the reduced clobbers?

3. However, what does seem to have an impact is removing the last (unused) argument of the command syscall function. It would be worth having various versions of the function, namely `syscall::command0` (no extra argument), `syscall::command1` (only `arg1`) and `syscall::command2` (both `arg1` and `arg2`). This would be useful in other drivers too, instead of writing a dummy zero value in the corresponding registers (costing move instructions and available registers for something else). This case is quite common, here are some examples:

    https://github.com/tock/libtock-rs/blob/54b7025b620d17703b424676ea7d8abdddc20a99/src/led.rs#L16-L18

    https://github.com/tock/libtock-rs/blob/54b7025b620d17703b424676ea7d8abdddc20a99/src/led.rs#L46-L50

    https://github.com/tock/libtock-rs/blob/1beae85f7e4f60043d094c59e959356a361213c5/src/console.rs#L63

    https://github.com/tock/libtock-rs/blob/baa48db4017621f1fc5d718a2c2d752b16d39e0f/src/timer.rs#L135

Reply:
> In that sense, it seems that the memory barrier is actually better: otherwise, an operation appearing after the debug syscall in the code but reordered before it by the compiler could mess up the manual debugging.

It took me a while to realize this, but you're right. If the compiler moves a faulting memory operation before the system call it could cause a misleading result (no debug print when one should have occurred). I think adding the memory clobber helps here, so I'll do so.

> 3\. However, what does seem to have an impact is removing the last (unused) argument of the command syscall function. It would be worth having various versions of the function, namely `syscall::command0` (no extra argument), `syscall::command1` (only `arg1`) and `syscall::command2` (both `arg1` and `arg2`).

I agree, and opened #102 to implement this. Once we have iterated on #102 , I'll come back and update this PR to use the new `command` system call.

Thank you for the thorough investigation!

Reply:
#102 was closed due to data leakage issues, but I implemented an alternative that seems to be agreed upon. I'll resolve this conversation.